perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java

```diff
@@ -0,0 +1,209 @@
+package cz.metacentrum.perun.core.provisioning;
+
+import cz.metacentrum.perun.core.api.Attribute;
+import cz.metacentrum.perun.core.api.Facility;
+import cz.metacentrum.perun.core.api.GenDataNode;
+import cz.metacentrum.perun.core.api.GenMemberDataNode;
+import cz.metacentrum.perun.core.api.Group;
+import cz.metacentrum.perun.core.api.HashedGenData;
+import cz.metacentrum.perun.core.api.Member;
+import cz.metacentrum.perun.core.api.Resource;
+import cz.metacentrum.perun.core.api.Service;
+import cz.metacentrum.perun.core.api.VosManager;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Generates data in format:
+ *
+ * attributes: {...hashes...}
+ * hierarchy: {
+ *    ** facility **
+ *    hashes: [...hashes...]
+ *    members: []
+ *    children: [
+ *      {
+ *        ** resource1 **
+ *        hashes: [...hashes...]
+ *        children: [
+ *          {
+ *            ** group A **
+ *            hashes: [...hashes...]
+ *            members: [...group members...]
+ *            children: [...sub groups...]
+ *          },
+ *          {
+ *            ** group B **
+ *            ...
+ *          }
+ *        ]
+ *        members: [
+ *          {
+ *            ** member 1 **
+ *            hashes: [...hashes...]
+ *          },
+ *          {
+ *            ** member 2 **
+ *            ...
+ *          }
+ *        ]
+ *      },
+ *      {
+ *        ** resource2 **
+ *        ...
+ *      }
+ *    ]
+ * }
+ *
+ * @author Vojtech Sassmann <vojtech.sassmann@gmail.com>
+ */
+public class GroupsHashedDataGenerator implements HashedDataGenerator {
+
+	private final PerunSessionImpl sess;
+	private final Service service;
+	private final Facility facility;
+	private final GenDataProvider dataProvider;
+	private final boolean filterExpiredMembers;
+
+	private GroupsHashedDataGenerator(PerunSessionImpl sess, Service service, Facility facility,
+	                                 boolean filterExpiredMembers) {
+		this.sess = sess;
+		this.service = service;
+		this.facility = facility;
+		this.filterExpiredMembers = filterExpiredMembers;
+		dataProvider = new GenDataProviderImpl(sess, service, facility);
+	}
+
+	@Override
+	public HashedGenData generateData() {
+		dataProvider.loadFacilityAttributes();
+
+		List<Resource> resources = sess.getPerunBl().getFacilitiesManagerBl().getAssignedResources(sess, facility);
+		resources.retainAll(sess.getPerunBl().getServicesManagerBl().getAssignedResources(sess, service));
```

Comment:
You can use FacilitiesManager.getAssignedResources(sess, facility, specificVo, specificService) instead and call it as:
```
List<Resource> resources = sess.getPerunBl().getFacilitiesManagerBl().getAssignedResources(sess, facility, null, service);
```

Reply:
Fixed.